### PR TITLE
Update all APIs to be mostly-compatible with GAX 4.0.0-alpha02

### PR DESCRIPTION
**Pull request by jskeet**
Most of these changes are hand-written, with the final giant commit being autogenerated.

We can periodically rebase this branch from main, regenerating right as the final commit each time to avoid conflicts.

It should compile and unit tests should pass, but I expect integration tests to fail at the moment for some APIs where we need to expose more via GAX - those changes all have FIXME in the text.

### CODE REVIEW

**Comment by jskeet** [comment]
Note: we expect "detect differences" to fail. There needs to be a tooling change there to pick up different TFMs to compare against.

File: apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SessionPoolManager.cs [hunk 1]
```diff
@@ -35,7 +37,13 @@ namespace Google.Cloud.Spanner.Data
     /// required.
     /// </summary>
     public sealed class SessionPoolManager
-    {
+    {        
+        // TODO: Should these be configurable?
```
Comment (amanda-tarafa) [inline apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SessionPoolManager.cs:41]: I wonder if there's a way for us to tie all of these to DI by taking advantage of the client builder.

At least offer a constructor for SessionPoolManager to receive a service provider etc.?

All of that of course can be done later.

**Review by amanda-tarafa — APPROVED** [review]
LGTM

I left a comment in one of the commits about maybe having a base client builder with more emulator support built in so we don't need to expose the adapters, etc. But since exposing the adapters is also useful for the channel affinity and some other use cases, I'm fine with just exposing what we need. Else we'll end up with many different base client builders for the many different use cases.

File: apis/Google.Cloud.PubSub.V1/Google.Cloud.PubSub.V1/PublisherServiceApiClientBuilder.cs [hunk 2]
```diff
@@ -81,10 +81,10 @@ internal async Task<ChannelBase> CreateChannelAsync(CancellationToken cancellati
         {
             // Note: no need to try to detect the channel pool here, as we know we don't want to use it.
             var effectiveBuilder = MaybeCreateEmulatorClientBuilder() ?? this;
-            var endpoint = effectiveBuilder.Endpoint ?? GetDefaultEndpoint();
+            var endpoint = effectiveBuilder.Endpoint ?? ServiceMetadata.DefaultEndpoint;
             var credentials = await effectiveBuilder.GetChannelCredentialsAsync(cancellationToken).ConfigureAwait(false);
-            var grpcAdapter = GrpcAdapter ?? DefaultGrpcAdapter;
-            return grpcAdapter.CreateChannel(endpoint, credentials, effectiveBuilder.GetChannelOptions());
+            // FIXME: we don't currently expose enough in GrpcAdapter or ClientBuilderBase to make this work.
```
[the remark this answers is not in the document]
Comment (jskeet) [reply apis/Google.Cloud.PubSub.V1/Google.Cloud.PubSub.V1/PublisherServiceApiClientBuilder.cs:86]: I'm not quite sure what we need to do yet - once this is in, I'll probably start on a GAX 4.0.0-alpha03 locally and see what comes out of it. We may need to expose the effective gRPC adapter, but *not* any other operations on it.

File: apis/Google.Cloud.Bigtable.V2/Google.Cloud.Bigtable.V2/BigtableServiceApiClientPartial.cs [hunk 3]
```diff
@@ -134,50 +130,32 @@ public partial class BigtableServiceApiSettings
...
     }
 
     public sealed partial class BigtableServiceApiClientBuilder : ClientBuilderBase<BigtableServiceApiClient>
     {
-        internal BigtableServiceApiClientBuilder(BigtableClientBuilder builder)
+        private static readonly GrpcChannelOptions s_bigtableDefaultOptions = DefaultOptions
+            .WithPrimaryUserAgent(BigtableClient.UserAgent)
+            // TODO: these were previously -1, which was treated as infinity by Grpc.Core, but is treated
+            // as "too small for any message" by Grpc.Net.Client. We could potentially hide that difference
+            // in the adapter instead of here, but using int.MaxValue feels more sensible anyway.
```
[the remark this answers is not in the document]
Comment (jskeet) [reply apis/Google.Cloud.Bigtable.V2/Google.Cloud.Bigtable.V2/BigtableServiceApiClientPartial.cs:153]: Yes, basically :)

File: apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SessionPoolManager.cs (hunk 1 above)
Comment (jskeet) [reply apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SessionPoolManager.cs:41]: Yes, that would probably make sense - and can be pushed until post GA, probably.

**Pull request merged**